From dd924e8bec816e54ab4dea1d3d4ecf2596998b8f Mon Sep 17 00:00:00 2001
From: fefifofum <fefifofum@email.com>
Date: Sun, 16 Nov 2014 17:27:41 +0100
Subject: [PATCH] Fix Xiaomi wifi mac address

---
 wifi/Android.mk |  5 +++++
 wifi/wifi.c     | 31 +++++++++++++++++++++++++++++++
 2 files changed, 36 insertions(+)

diff --git a/wifi/Android.mk b/wifi/Android.mk
index c7bdc59..80dfe76 100644
--- a/wifi/Android.mk
+++ b/wifi/Android.mk
@@ -46,4 +46,9 @@ ifeq ($(BOARD_HAVE_SAMSUNG_WIFI),true)
 LOCAL_CFLAGS += -DSAMSUNG_WIFI
 endif
 
+ifeq ($(BOARD_HAVE_XIAOMI_WIFI),true)
+LOCAL_CFLAGS += -DXIAOMI_WIFI
+LOCAL_SHARED_LIBRARIES += libqminvapi
+endif
+
 LOCAL_SHARED_LIBRARIES += libnetutils
diff --git a/wifi/wifi.c b/wifi/wifi.c
index 80a0e2a..a0dd5c6 100644
--- a/wifi/wifi.c
+++ b/wifi/wifi.c
@@ -149,6 +149,13 @@ static char supplicant_name[PROPERTY_VALUE_MAX];
 static char supplicant_prop_name[PROPERTY_KEY_MAX];
 
 
+#ifdef XIAOMI_WIFI
+extern int qmi_nv_read_wlan_mac(char** mac);
+static unsigned char wlan_addr[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, };
+static char wcn_mac_arg[120];
+static int read_mac_ok = 0;
+#endif
+
 #ifdef SAMSUNG_WIFI
 char* get_samsung_wifi_type()
 {
@@ -298,8 +305,14 @@ int wifi_load_driver()
     usleep(200000);
 #endif
 
+#ifdef XIAOMI_WIFI
+    if (0 == read_mac_ok)
+        read_wlan_mac_addr();
+    if (insmod(DRIVER_MODULE_PATH, wcn_mac_arg) < 0) {
+#else
     if (insmod(DRIVER_MODULE_PATH, DRIVER_MODULE_ARG) < 0) {
 #endif
+#endif
 
 #ifdef WIFI_EXT_MODULE_NAME
         rmmod(EXT_MODULE_NAME);
@@ -1166,3 +1179,21 @@ int wifi_set_mode(int mode) {
     wifi_mode = mode;
     return 0;
 }
+
+#ifdef XIAOMI_WIFI
+int read_wlan_mac_addr()
+{
+    char* nv_wlan_mac = NULL;
+    ALOGV("read wlan mac addr enter.");
+    qmi_nv_read_wlan_mac(&nv_wlan_mac);
+    int i;
+    for (i=0; i<6; i++) {
+        wlan_addr[i] = nv_wlan_mac[i];
+    }
+    memset(wcn_mac_arg, 0, sizeof(wcn_mac_arg));
+    sprintf(wcn_mac_arg, "mac=0x%02x,0x%02x,0x%02x,0x%02x,0x%02x,0x%02x", wlan_addr[5], wlan_addr[4], wlan_addr[3], wlan_addr[2],wlan_addr[1],wlan_addr[0]);
+    read_mac_ok = 1;
+    ALOGV("read wlan mac addr (%s) done.", wcn_mac_arg);
+    return 0;
+}
+#endif
-- 
1.9.1

